src/components/product-progress/product-progress.repository.ts

```diff
@@ -1,7 +1,189 @@
+        }),
+        node('node', 'ProductProgress'),
+        relation('in', 'reportProgressRel', 'reportProgress', { active: true }),
+        node('report', 'PeriodicReport', { id: reportId }),
+      ])
+      .setValues({
+        'productProgressRel.active': false,
+        'reportProgressRel.active': false,
+      })
+      .return('node.id as id')
+      .asResult<{ id: ID }>()
+      .run();
+  }
+
+  async create(input: ProductProgressInput) {
+    const id = await generateId();
+    const createdAt = DateTime.local();
+    const query = this.db
+      .query()
+      .match([node('product', 'Product', { id: input.productId })])
+      .match([node('report', 'PeriodicReport', { id: input.reportId })])
+      .create([
+        [
+          node('node', ['ProductProgress', 'BaseNode'], {
+            id,
+            createdAt,
+          }),
+        ],
+      ])
+      .create([
+        node('product'),
+        relation('out', '', 'productProgress', { active: true, createdAt }),
+        node('node'),
+        relation('in', '', 'reportProgress', { active: true, createdAt }),
+        node('report'),
+      ])
+      .return('node.id as id')
+      .asResult<{ id: ID }>();
+    const result = await query.first();
+
+    if (!result?.id) {
+      return;
+    }
+    await Promise.all(
+      input.steps.map(async (step) => {
+        const id = await generateId();
+        const createdAt = DateTime.local();
+        const query = this.db
+          .query()
+          .match([
+            node('productProgress', 'ProductProgress', { id: result.id }),
+          ])
+          .create([
+            [
+              node('node', ['StepProgress', 'BaseNode'], {
+                id,
+                createdAt,
+              }),
+            ],
+            ...property('step', step.step, 'node'),
+            ...property('percentDone', step.percentDone, 'node'),
+            ...property('description', step.description, 'node'),
+          ])
+          .create([
+            node('productProgress'),
+            relation('out', '', 'step', { active: true, createdAt }),
+            node('node'),
+          ])
+          .return('node.id as id')
+          .asResult<{ id: ID }>();
+
+        await query.first();
+      })
+    );
+  }
+
+  async readSteps(productId: ID, reportId: ID) {
+    const result = await this.db
+      .query()
+      .match([
+        node('product', 'Product', { id: productId }),
+        relation('out', '', 'productProgress', { active: true }),
+        node('node', 'ProductProgress'),
+        relation('in', '', 'reportProgress', { active: true }),
+        node('report', 'PeriodicReport', { id: reportId }),
+      ])
+      .match([
+        node('node'),
+        relation('out', '', 'step', { active: true }),
+        node('stepProgress', 'StepProgress'),
+      ])
+      .with('collect(stepProgress.id) as ids')
+      .return('ids')
+      .asResult<{ ids: [ID] }>()
+      .first();
```

Comment:
This call `this.hydrate()` here instead of returning only the IDs. Takes this query from O(n+1) to O(1)